src/vcf/sqlite_report.cpp

```diff
@@ -0,0 +1,310 @@
+/**
+ * Copyright 2016 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "vcf/sqlite_report.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+    SqliteReportRW::SqliteReportRW(std::string db_name) : db_name{db_name}, current_transaction_size{0},
+                                                          transaction_size{1000000}, sleep_time{500}
+    {
+        int rc = sqlite3_open(db_name.c_str(), &db);
+        if (rc != SQLITE_OK) {
+            sqlite3_close(db);
+            throw std::runtime_error(std::string{"Can't open database: "} + sqlite3_errmsg(db));
+        }
+
+        char *zErrMsg = NULL;
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists errors (line int, code int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string{"Can't use table: "} + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists warnings (line int, code int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string{"Can't use table: "} + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+    }
+
+    void SqliteReportRW::flush()
+    {
+        // in case the amount of inserts is not a multiple of the transaction size, do the leftover inserts in a smaller transaction
+        if (current_transaction_size != 0) {
+            commit_transaction();
+            current_transaction_size = 0;
+        }
+    }
+    
+    void SqliteReportRW::close()
+    {
+        flush();
+        
+        if (db != nullptr) {
+            int rc = sqlite3_close(db);
+            if (rc != SQLITE_OK) {
+                int remaining_tries = 10;
```

Comment:
I made it the more intuitive way I could. I am using a postdecrement, in the next condition, so a `remaining_tries = 10` means that it will sleep 10 times. in the last iteration, it is doing:
```
// remaining_tries = 1;
while (... remaining_tries-- > 0){   // 1 > 0 true
    // remaining_tries = 0;
    ... // last wait
}
while (... remaining_tries-- > 0){ // 0 > 0 false
} // exit loop
// remaining_tries = -1; // but it is not used anymore
```
using `unsigned` leads either a less intuitive loop of less intuitive counts. or we can use a `for` loop of a condition `rc == SQLITE_BUSY && i < remaining_tries`